### PR TITLE
Optional filtering blacklist regions

### PR DESCRIPTION
**Pull request by orzechoj**
Added a process to filter out reads in blacklist regions, given a parameter "blacklist" pointing to a bed file. The default option is to skip this step.

### CODE REVIEW

File: main.nf [hunk 1]
```diff
@@ -287,24 +289,51 @@ process bwa {
...
  */
 
+/* Optional filtering of blacklist regions. */
+
+if(!params.blacklist){
+	filtered_bam = bwa_bam
+} else {
+  process bedtools_intersect {
+    	  tag "${bam.baseName}"
+
```
Comment (ewels) [inline main.nf:299]: Please add `publishDir` here:

```groovy
publishDir path: { params.saveAlignedIntermediates ? "${params.outdir}/filter_blacklist" : params.outdir }, mode: 'copy',
           saveAs: {filename -> params.saveAlignedIntermediates ? filename : null }
```

The above code will only save these files when `--saveAlignedIntermediates` is specified.

Does `bedtools intersect` create any log files or stdout? Would be good to save this if so.

Comment (ewels) [reply main.nf:299]: ps. If saving log output, please put it in a subdirectory as follows:

```groovy
publishDir path: { params.saveAlignedIntermediates ? "${params.outdir}/filter_blacklist" : params.outdir }, mode: 'copy',
           saveAs: {filename ->
                if (filename.indexOf(".bam") == -1) "logs/$filename"
                else params.saveAlignedIntermediates ? filename : null
           }
```

**Review by ewels — APPROVED** [review]
Looks great, thanks!

**Comment by ewels** [comment]
Test failures are due to a unrelated problem with the docker image I think. I thought we'd fixed these already? 😞

**Comment by chuan-wang** [comment]
Yes, please merge my PR first then it will work.

**Comment by ewels** [comment]
Now somehow have a merge conflict :(

**Comment by chuan-wang** [comment]
Hi Jakub (@orzechoj) and Phil (@ewels),

I had another thought about this PR on my way back home, and realized two issues that might be worthy of further discussions . Please correct me if I have misunderstood :D

1) Removing sequence reads from bam files works perfectly for single-end datasets. However, for paired-end datasets ChIP-seq peaks are called by shifting the reads aligned to forward and reverse strands,, which means that the original coordinates of sequence reads do not really overlap with the blacklisted regions.
2) The statistics such as fingerprints and NSC/RSC values may be skewed since sequence reads from blacklisted regions are artificially removed.

In this sense I guess it might be better to remove ChIP-seq peaks which overlap with blacklisted regions after peak calling. We can discuss it tomorrow if you have time.

Best,
Chuan

File: main.nf (hunk 1 above)
Comment (orzechoj) [reply main.nf:299]: Thanks. There's no log output from this step.

**Comment by chuan-wang** [comment]
Hi Phil (@ewels), Jakub and I just had a discussion about this PR. We agreed that now we can keep it on hold. I will make a new PR which includes blacklisted region filtering after peak calling. I will also add new functions such as peak annotations. /chuan

**Comment by orzechoj** [comment]
I just talked to Agata Smialowska at NBIS who has been doing at lot of ChIP-seq things. She commented that for single read data, it was good to filter out the reads in the blacklist regions early, so these reads don’t confuse the fragment length calculations (phantompeaktools). But in most cases I don’t think it matters when you remove the blacklist regions.

Anyway, I’ll keep my version, perhaps it will be useful to compare the the two approaches (removing reads in blacklist regions early or removing peaks in blacklist regions) when I get the ChIP-seq data I’m waring for.

**Comment by chuan-wang** [comment]
It would be nice to compare these two strategies for filtering peaks!

**Comment by ewels** [comment]
Yes, my gut feeling is that it's good to remove the reads as early as possible - these are not true alignments, these fragments of DNA don't really come from these regions in the genome, so any NSC / RSC / other stats calculated with them _in place_ will be skewed (rather than the other way around).

However, as discussed with @orzechoj I can't imagine that it makes a lot of difference, and it's nice to do the same thing as most other people (eg. encode), so I'm happy either way.

@orzechoj - I'll close this PR without merging now, but it would be great if you could move these commits onto their own branch so that we can still run the code and compare later 👍 

Phil

**Pull request closed without merging**